File: src/Solution_0818.java
import java.util.HashSet;

public class Solution_0818 {
    public int lengthOfLongestSubString1(String s) {
        int len = s.length();
        if (len <= 1) {
            return len;
        }
        int start = 0, end = 0, res = 0;
        while (end < len) {
            end++;
            while (end < len && s.charAt(end) == s.charAt(start)) {
                res = Math.max(end - start, res);
                start++;
                end = start;
            }
            if (res == len) {
                return res;
            }
        }
        return res;
    }

    public int lengthOfLongestSubString(String s) {
        int len = s.length();
        if (len <= 1) {
            return len;
        }
        HashSet<Character> set = new HashSet<>();
        int start = 0, end = 0, res = 0;
        while (end < len) {
            while (end < len && !set.contains(s.charAt(end))) {
                set.add(s.charAt(end));
                end++;
                res = Math.max(res, end - start);
            }
            set.remove(s.charAt(start));
            start++;
        }
        return res;
    }

    public static void main(String[] args) {
        Solution_0818 solution = new Solution_0818();
        String s = "pwwkew";
        int result = solution.lengthOfLongestSubString(s);
        System.out.println(result);
    }
}

File: src/Solution_0818.js
/**
 *
 * @param {String} s
 * @return {number}
 */
var lengthOfLongestSubString = function(s){
    const len = s.length;
    let start = 0, end = 0, res = 0;
    const set =  new Set();
    while(end < len){
        while(end < len && !set.has(s[end])){
            set.add(s[end]);
            end ++;
            res = Math.max(res, end - start);
        }
        // set里已经有这个值；
        set.delete(s[start]);
        start ++;
    }
    return res;
}

const s = lengthOfLongestSubString("pwwkew");
console.log(s);

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function (head) {
    const set = new Set();
    while (head !== null) {
        const prevSize = set.size;
        set.add(head);
        const curSize = set.size;
        if (curSize === prevSize) {
            return true;
        }
        head = head.next;
    }
    return false;
};
